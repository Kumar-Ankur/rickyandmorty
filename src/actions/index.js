import * as actionTypes from "../types";

/**
 * @function getCharacters
 * @description action types for fetching characters data from API using saga file and update the state.
 */
export function getCharacters() {
  return { type: actionTypes.GET_CHARACTERS };
}

/**
 * @function getFilteredData
 * @param { Object } data - Filtered data
 * @description - filtered data based on payload value.
 */
export function getFilteredData(data) {
  return { type: actionTypes.GET_FILTERED_DATA, payload: data };
}

/**
 * @function sortAscendingById
 * @description - sort all the character data in ascending order
 */
export function sortAscendingById() {
  return { type: actionTypes.SORT_DATA_BY_ASCENDING };
}

/**
 * @function sortDescendingById
 * @description - sort all the character data in descending order.
 */
export function sortDescendingById() {
  return { type: actionTypes.SORT_DATA_BY_DESCENDING };
}

/**
 * @function addFilteredData
 * @param { String } data
 * @description - add filtered data in string for species category state
 */
export function addFilteredData(data) {
  return { type: actionTypes.ADD_FILTERED_DATA, payload: data };
}

/**
 * @function removeFilteredData
 * @param { String } data
 * @description - remove filtered data in string for species category state
 */
export function removeFilteredData(data) {
  return { type: actionTypes.REMOVE_FILTERED_DATA, payload: data };
}

/**
 * @function addGenderFilteredData
 * @param { String } data
 * @description - add filtered data in string for gender category state
 */
export function addGenderFilteredData(data) {
  return { type: actionTypes.ADD_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function removeGenderFilteredData
 * @param { String } data
 * @description - remove filtered data in string for gender category state
 */
export function removeGenderFilteredData(data) {
  return { type: actionTypes.REMOVE_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function addOriginFilteredData
 * @param { String } data
 * @description - add filtered data in string for origin category state
 */
export function addOriginFilteredData(data) {
  return { type: actionTypes.ADD_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function removeOriginFilteredData
 * @param { String } data
 * @description - remove filtered data in string for origin category state
 */
export function removeOriginFilteredData(data) {
  return { type: actionTypes.REMOVE_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function addSearchText
 * @param { String } data
 * @description - add search data in string for search category state
 */
export function addSearchText(data) {
  return { type: actionTypes.ADD_SEARCH_TEXT, payload: data };
}
