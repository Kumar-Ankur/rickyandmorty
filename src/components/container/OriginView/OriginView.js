import React, { Component } from "react";
import CheckBox from "../../presentation/CheckBox/CheckBox";
import { connect } from "react-redux";
import * as actionCreator from "../../../actions";
import * as constants from "../../../constants";
import { getCheckBoxData } from '../../../utils';
import './OriginView.scss';

class OriginView extends Component {

  render() {
    const { all_characters } = this.props;
    const originData = getCheckBoxData(all_characters, constants.ORIGIN, true);
    const { addOriginFilteredData, removeOriginFilteredData } = this.props;
    const getOriginBox = originData.map((data, index) => {
      return (
        <div key={index}>
          <CheckBox
            data={data}
            addFilter={addOriginFilteredData}
            removeFilter={removeOriginFilteredData}
          />
        </div>
      );
    });

    return (
      <div className="species-content">
          <h2>{ constants.ORIGIN_FILTER }</h2>
        <form>{getOriginBox}</form>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    all_characters: state.character.all_characters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    addOriginFilteredData: (data) => dispatch(actionCreator.addOriginFilteredData(data)),
    removeOriginFilteredData: (data) =>
      dispatch(actionCreator.removeOriginFilteredData(data)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(OriginView);
