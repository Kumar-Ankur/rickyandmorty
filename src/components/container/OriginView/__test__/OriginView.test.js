import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "../../../../reducers";
import OriginView from "../OriginView";
import { mapDispatchToProps } from "../OriginView";
import * as actionCreator from "../../../../actions";

const middleware = createSagaMiddleware();
const mockStore = createStore(rootReducer, applyMiddleware(middleware));

/**
 * Factory function to create a shallowwrapper for the App component.
 * @function setUp
 * @param { Object } props - Component props specific to this setup
 * @param { Object } state - Initial state for setUp
 * @returns { shallowWrapper }
 *
 */
function setUp(props = {}) {
  const wrapper = shallow(<OriginView store={mockStore} {...props} />);
  return wrapper;
}

/**
 * Return shallowWarpper containing node(s) with the given data-test attribute
 * @function findByTestAttr
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper to search within
 * @param {string} val - value of data-test attribute for search
 * @returns { shallowWrapper}
 */
function findByTestAttr(wrapper, val) {
  return wrapper.find(`[data-test="${val}"]`);
}

describe("check origin component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = setUp().dive().dive();
  });

  test("generate snapshot for origin component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("check originView component render without crash", () => {
    const componentOrigin = findByTestAttr(wrapper, "component-origin");
    expect(componentOrigin.length).toBe(1);
  });

  test("check origin checkbox form render without crash", () => {
    const checkbox = findByTestAttr(wrapper, "component-origin-form");
    expect(checkbox.length).toBe(1);
  });

  test("check `addOriginFilteredData` function inside mapDispatchToProps working fine", () => {
    const dispatchSpy = sinon.spy();
    mockStore.dispatch = dispatchSpy;
    const { addOriginFilteredData } = mapDispatchToProps(dispatchSpy);
    addOriginFilteredData("Earth");
    const lastSpyCall = dispatchSpy.args[0][0];
    const expectedAction = actionCreator.addOriginFilteredData("Earth");

    expect(lastSpyCall.types).toEqual(expectedAction.types);
  });

  test("check `removeOriginFilteredData` function inside mapDispatchToProps working fine", () => {
    const dispatchSpy = sinon.spy();
    mockStore.dispatch = dispatchSpy;
    const { removeOriginFilteredData } = mapDispatchToProps(dispatchSpy);
    removeOriginFilteredData("Earth");
    const lastSpyCall = dispatchSpy.args[0][0];
    const expectedAction = actionCreator.removeOriginFilteredData("Earth");

    expect(lastSpyCall.types).toEqual(expectedAction.types);
  });
});
