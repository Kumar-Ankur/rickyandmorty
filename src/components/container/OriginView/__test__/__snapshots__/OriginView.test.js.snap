// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check origin component generate snapshot for origin component 1`] = `ShallowWrapper {}`;
